### PR TITLE
Dependency updates for v2024.4

### PR DESCRIPTION
**Pull request by grzesiek2010**
#### Why is this the best possible solution? Were any other approaches considered?
As usual, I've updated the dependencies for the next release. However, this time, there are some significant changes in addition to the regular updates:

- **Switched to Gradle Version Catalog**: Dependency management now uses the [Gradle Version Catalog](https://developer.android.com/build/migrate-to-catalogs), which is the default approach for managing dependencies in new projects. It offers several advantages over the previous `buildSrc` module, such as faster project builds when dependencies are changed and notifications about new available versions. For details, see [this commit](https://github.com/getodk/collect/pull/6505/commits/4016107e709fee55e0975a2a3082634d78a5e9b0).
- **Migrated to FusedLocationProviderClient**: This migration was necessary to update the `play-services-location` library. For more information, see [this commit](https://github.com/getodk/collect/pull/6505/commits/9ce5c4cc5e502ed34d8907eb28d65dff847daa36).
- **Reviewed and Fixed Some PMD Rules**: Over the last few dependency updates, we had disabled some new PMD rules. After a review, I re-enabled a few of them. However, others were left disabled as they either don’t make much sense, would require extensive edits across hundreds of classes, or are faulty and don’t work as intended.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It's often challenging to determine what should be tested when multiple dependencies are updated. However, in this case, I can highlight one area: location recording. As mentioned earlier, we switched to a new location provider, which could be a potential source of bugs. In fact, this is our second attempt at the migration, as the first one failed due to this issue: https://github.com/getodk/collect/issues/6027. Hopefully, everything will go smoothly this time.

#### Do we need any specific form for testing your changes? If so, please attach one.
A form with a few geopints might be helpful for testing reading locations: 
[geo5.xlsx](https://github.com/user-attachments/files/17815778/geo5.xlsx)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: gradle/libs.versions.toml
Comment (seadowg) [inline gradle/libs.versions.toml:101]: Is there a way to declare the plugin versions here so we don't still need these declared in the root `build.gradle`?

File: async/build.gradle.kts
Comment (seadowg) [inline async/build.gradle.kts:9]: I'm thinking it might make sense to still declare the compile/target/mid SDKs in as constants in Gradle (example below). It doesn't seem like version catalog lets us define integer values. They're also more like shared configuration than library versions.

```kotlin
buildscript {
    ext {
        minSdkVersion = 21
        targetSdkVersion = 34
        compileSdkVersion = 34
    }
...
```

File: location/src/main/java/org/odk/collect/location/GoogleFusedLocationClient.kt
Comment (seadowg) [inline location/src/main/java/org/odk/collect/location/GoogleFusedLocationClient.kt:23]: Did any changes get made to `GoogleFusedLocationClient` or its test? Because the Kotlin version happens in the same commit, the diff doesn't give me any insight into that.

**Review by seadowg — CHANGES_REQUESTED** [review]
Loving the change to version catalog! I used it recently on a different app and thought it was so much nicer, but definitely a pain to convert to. Thanks for taking the time to do that.

File: async/build.gradle.kts
Comment (grzesiek2010) [reply async/build.gradle.kts:9]: In kotlin DSL this would be like this:
```
    compileSdk = rootProject.extra["compileSdkVersion"] as Int
```

so it would also require casting to Int. Given that I don't think this approach would be any better.

File: location/src/main/java/org/odk/collect/location/GoogleFusedLocationClient.kt
Comment (grzesiek2010) [reply location/src/main/java/org/odk/collect/location/GoogleFusedLocationClient.kt:23]: Ah, yes, there are some changes because the old provider used different listeners, so the file structure has changed slightly. The file isn't large, so hopefully comparing it with the old one shouldn't be difficult.

File: gradle/libs.versions.toml
Comment (grzesiek2010) [reply gradle/libs.versions.toml:101]: Yes, I even attempted to do that, but since it was quite complex (requiring a lot of additional changes) and some of the functions I would need to use are still marked as unstable, I decided to hold off for now.

File: async/build.gradle.kts
Comment (seadowg) [reply async/build.gradle.kts:9]: Oh, wow. Kotlin makes that much worse - in Groovy it'd be `rootProject.ext.minSdkVersion`. Let's leave it as-is then.

**Review by seadowg — APPROVED** [review]

**Comment by seadowg** [comment]
@getodk/testers I think it's worth testing location tracking (background, maps etc) before merging this as we've made changes to how we interact with location services.

**Comment by dbemke** [comment]
@grzesiek2010 Could you send us (in Slack) the apk with access to different maps sources?

**Comment by dbemke** [comment]
The first map opening in Geotrace or Geoshape feels a bit slower than it used to be. It seems that it happens after installing the app and opening the map for the first time. For example, on Android 10 (Redmi 9t) in OSM it takes 10 seconds before the maps zooms in / finishes searching for location. It works well but seems a bit slower.

**Comment by grzesiek2010** [comment]
Hmm I've just tested it and it took 3s (fresh install) for all providers (Google, Mapbox, OSM). @srujner @WKobus are you also experiencing such differences between versions?

**Comment by WKobus** [comment]
I've tried it now on different providers and I had same results every time it took 2-3s, however I remember that during testing this PR I had a situation when it took ~10 sec to load, but I can't reproduce it anymore.

**Comment by srujner** [comment]
@grzesiek2010 Same with me, during testing I encountered this issue once or twice but now I have not been able to reproduce it anymore.

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android: 5.1, 8.1, 12, 14 

**Verified cases:**

    • Collecting Location in background: ; 
    • Viewing Submissions on map; 
    • Regression checks for GeoPoint, GeoTrace, GeoShape widgets; 
    • Select Map forms 
    • Offline Layers; 
    • Google Maps, Mapbox, OSM; 
    • Collecting Location in background; 
    • View Submissions on map; 
    • Screen rotation, minimize app, using backward button, 
    • Forms from issue

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android:  8.1, 10, 14

**Comment by WKobus** [comment]
Tested with Success!

Verified on device with Android: 11, 15

**Comment by grzesiek2010** [comment]
> I've tried it now on different providers and I had same results every time it took 2-3s, however I remember that during testing this PR I had a situation when it took ~10 sec to load, but I can't reproduce it anymore.

> Same with me, during testing I encountered this issue once or twice but now I have not been able to reproduce it anymore.

So, it might have been random. Sometimes a device needs more time to read the location, and it doesn’t necessarily have to be related to these changes. Let’s keep an eye on it.

**Pull request merged**
